<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

**Exercise 1 (Make me an odd list)**

With your knowledge of growing lists, use a for loop with range() and continue to generate a list of the squares of the odd integers from 0 to 9.

Hint: You can check for ‘evenness’ using number % 2 == 0.

In [1]:
%%timeit 

for number in range(10):
    if number % 2 == 0:
        continue
    number**2



1.07 μs ± 305 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


**Exercise 2 (Make me another odd list)**

In [2]:
%%timeit

[number**2 for number in range(10) if number % 2 != 0]


903 ns ± 21 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


**Exercise 3 (Time me!)**

List comprehension is faster by about 1.2x.

1.07 μs ± 305 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
903 ns ± 21 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

**Exercise 4 (A problem of decay)**
The initial quantity of a sample of a radioactive substance is 100 units, and it decays by 5% each year. Use a while loop to determine how long the sample will take to reduce to half its original amount.

In [8]:
substance = 100
count = 0

while substance > 50:
    substance = substance*0.95
    count += 1


print(f'Years: {count}, Substance left: {substance:.2f} units')



Years: 14, Substance left: 48.77 units


**Exercise 5 (Changes in CO2)**
Identify those years that showed an increase of CO
 of 3 ppm or more compared to the previous year.
Please print out these years along with the corresponding change in concentration.

In [9]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]


In [32]:
#for year, co2 in co2_data:

import numpy as np
co2_data = np.array(co2_data)


#for count, dataset in enumerate(co2_data):
    #print(f'{count}: {dataset}')
    #if dataset[i+1] 


#enumerate, then use the enumerated number as the "key" to compare the dataset
    

only_co2 = co2_data[:,1]
print(only_co2)

for i in range(len(only_co2)):
    if only_co2[i+1] - only_co2[i] > 3:
        print(i)
#sliced the list into only the co2, wanted to use index of the list to do it but cannot?


#for year, co2 in co2_data:
    #if 



[369.55 371.14 373.28 375.8  377.52 379.8  381.9  383.79 385.6  387.43
 389.9  391.65 393.85 396.52 398.65 400.83 404.24 406.55 408.52 411.44
 414.24]
15


IndexError: index 21 is out of bounds for axis 0 with size 21